plugins/woocommerce/includes/class-wc-ajax.php

```diff
@@ -3121,6 +3124,22 @@ public static function toggle_gateway_enabled() {
 		wp_send_json_error( 'invalid_gateway_id' );
 		wp_die();
 	}
+
+	/**
+	 * Reimplementation of WP core's `wp_ajax_add_meta` method to support order custom meta updates with custom tables.
+	 */
+	public static function order_add_meta() {
```

Comment:
Note that these methods are only placeholders and functionality is actually provided by internal classes that these methods are immediately calling. I didn't want to add another internal Ajax controller to keep ajax related logic in one place (headers, security etc). I can add a comment perhaps saying that these methods are just proxy to internal methods, and no business logic should be added directly

Reply:
> I didn't want to add another internal Ajax controller to keep ajax related logic in one place

In this case the proper thing to do (by "proper" here I mean according to our guidelines/coding practices) is to add any new methods to a new class in `src` and, if necessary, move existing functionality from the old class to the new one, turning the old methods into stubs to methods in the new class if necessary.

That said, I see that these two methods are also added to the `$ajax_events` array and thus are direct targets for ajax calls, therefore actually moving them to a different class would require a significant refactor that is probably out of scope for this pull request.

Counter-that said, having to add new public methods just for the sake of being the target of hooks is a problem that is in principle solved by https://github.com/woocommerce/woocommerce/pull/34019, but unfortunately in this case the target methods are static and thus that solution can't be used. So while we find a solution for that, could we use the old "function calling to private methods" strategy here, even if temporarily? It would be a matter of making the methods private, removing `order_add_meta` and `order_delete_meta` from `$ajax_events`, and adding something like the following at the end of `add_ajax_events`:

```php
$ajax_events = array(
	'order_add_meta',
	'order_delete_meta',
);

foreach ( $ajax_events as $ajax_event ) {
	add_action( 'wp_ajax_woocommerce_' . $ajax_event,
		function() use ( $ajax_event ) {
			call_user_func( array( __CLASS__, $ajax_event ) );
		}
	);
}
```

Reply:
makes sense, addressed in https://github.com/woocommerce/woocommerce/pull/33972/commits/de890c55a350185f8a36f9731b7ebf9e034d8832